pkg/config/v2/template/registry.go

```diff
@@ -0,0 +1,194 @@
+// @license
+// Copyright 2021 Dynatrace LLC
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package template
+
+import (
+	"encoding/json"
+	"path/filepath"
+
+	templ "text/template"
+
+	"github.com/spf13/afero"
+)
+
+type Template interface {
+	// id of the template, used as a key in the template cache
+	Id() string
+
+	// human readable name for this template, mostly used for debugging
+	Name() string
+
+	// string content of the template
+	Content() string
+}
+
+type FileBasedTemplate interface {
+	Template
+	FilePath() string
+}
+
+// type defining a template which can be rendered
+type stringTemplate struct {
+	id string
+
+	name string
+
+	content string
+}
+
+func (t *stringTemplate) Id() string {
+	return t.id
+}
+
+func (t *stringTemplate) Name() string {
+	return t.name
+}
+
+func (t *stringTemplate) Content() string {
+	return t.content
+}
+
+// type defining a template which can be rendered
+type fileBasedTemplate struct {
+	path    string
+	content string
+}
+
+func (t *fileBasedTemplate) Id() string {
+	return t.path
+}
+
+func (t *fileBasedTemplate) Name() string {
+	return t.path
+}
+
+func (t *fileBasedTemplate) Content() string {
+	return t.content
+}
+
+func (t *fileBasedTemplate) FilePath() string {
+	return t.path
+}
+
+var (
+	_ Template          = (*stringTemplate)(nil)
+	_ FileBasedTemplate = (*fileBasedTemplate)(nil)
+)
+
+func (t *stringTemplate) MarshalJSON() ([]byte, error) {
+	// Only used for debugging purpose.
+	// The content of the template is dropped, since it adds too much
+	// content to the debug file.
+	return json.Marshal(&struct{ Name string }{Name: t.name})
+}
+
+// cache for templates so that they don't get read from disk multiple times
+var templateCache map[string]Template = make(map[string]Template)
```

Comment:
No need for type here
```suggestion
var templateCache = make(map[string]Template)
```